src/lib/pubSub/index.ts

```diff
@@ -1,43 +1,50 @@
-// https://davidwalsh.name/pubsub-javascript
 // we use pubsub to sync events between multiple components(example - ChannelList, Channel)
 // for example, if customer sends a message from their custom component
 // without pubsub,we would not be able to listen to it
 // in our ChannelList or Conversation
 
-export interface PubSubTypes {
-  __getTopics: () => Record<string, string>;
-  subscribe: (topic: string, listener: unknown) => { remove: () => void };
-  publish: (topic: string, info: unknown) => void;
+export type PubSubTypes<
+  T extends keyof any,
+  U extends { topic: T; payload: any }
+> = {
+  __getTopics: () => Record<string, Set<(data: PayloadByTopic<T, U>) => void>>;
+  subscribe: <K extends T>(
+    topic: K,
+    subscriber: (data: PayloadByTopic<K, U>) => void
+  ) => { remove: () => void };
+  publish: <K extends T>(topic: K, data: PayloadByTopic<K, U>) => void;
+};
+
+type PayloadByTopic<T extends keyof any, U> = U extends {
+  topic: T;
+  payload: infer P;
 }
+  ? P
+  : never;
 
-const pubSubFactory = (): PubSubTypes => {
-  const topics = {};
-  const hOP = topics.hasOwnProperty;
+const pubSubFactory = <
+  T extends keyof any = keyof any,
+  U extends { topic: T; payload: any } = any
```

Comment:
nice but why we need default values here?
```
= keyof any,
= any
```

Reply:
for pubSubFactory testing